pkg/controllers/provisioning/scheduling/topologygroup.go

```diff
@@ -69,20 +69,42 @@ type TopologyGroup struct {
 	emptyDomains sets.Set[string]       // domains for which we know that no pod exists
 }
 
-func NewTopologyGroup(topologyType TopologyType, topologyKey string, pod *v1.Pod, namespaces sets.Set[string], labelSelector *metav1.LabelSelector, maxSkew int32, minDomains *int32, domains sets.Set[string]) *TopologyGroup {
-	domainCounts := map[string]int32{}
-	for domain := range domains {
-		domainCounts[domain] = 0
-	}
+func NewTopologyGroup(topologyType TopologyType, topologyKey string, pod *v1.Pod, namespaces sets.Set[string], labelSelector *metav1.LabelSelector, maxSkew int32, minDomains *int32, taintPolicy *v1.NodeInclusionPolicy, affinityPolicy *v1.NodeInclusionPolicy, domainGroup TopologyDomainGroup) *TopologyGroup {
 	// the nil *TopologyNodeFilter always passes which is what we need for affinity/anti-affinity
 	var nodeSelector TopologyNodeFilter
 	if topologyType == TopologyTypeSpread {
-		nodeSelector = MakeTopologyNodeFilter(pod)
+		nodeTaintsPolicy := v1.NodeInclusionPolicyIgnore
+		if taintPolicy != nil {
+			nodeTaintsPolicy = *taintPolicy
+		}
+		nodeAffinityPolicy := v1.NodeInclusionPolicyHonor
+		if affinityPolicy != nil {
+			nodeAffinityPolicy = *affinityPolicy
+		}
+		nodeSelector = MakeTopologyNodeFilter(pod, nodeTaintsPolicy, nodeAffinityPolicy)
 	}
+
 	selector, err := metav1.LabelSelectorAsSelector(labelSelector)
 	if err != nil {
 		selector = labels.Nothing()
 	}
+
+	domains := map[string]int32{}
+	if nodeSelector.TaintPolicy == v1.NodeInclusionPolicyHonor {
+		domainGroup.ForEachToleratedDomain(pod, func(domain string) {
+			domains[domain] = 0
+		})
+	} else {
+		domainGroup.ForEachDomain(func(domain string) {
+			domains[domain] = 0
+		})
+	}
+
+	emptyDomains := sets.New[string]()
```

Comment:
So, does emptyDomains need to match the domains that are marked above? If a domain isn't tolerated, is it going to be problematic if we allow it to be an empty domain -- if not practically, what about semantically?

Reply:
Discussed offline, while this makes some sense since `emptyDomains` is only used for affinity based topology groups, which don't support an honor taints policy, mapping `emptyDomains` directly to the keys of `domains` does make the intention clearer. It shouldn't make a functional difference, so we'll take in the change.

Reply:
Maybe just a personal preference thing but it makes more sense in my head to just inject into empty domains right next to where you are setting the value to 0 (which means you will have two lines vs. one) but I think it's worth it just to reason about the logic that's mapped together

Reply:
Nice! This looks great!